src/webgpu/shader/validation/uniformity/uniformity.spec.ts

```diff
@@ -356,6 +356,246 @@ fn main(@builtin(${t.params.builtin}) p : ${t.params.type}) {
     t.expectCompileResult(t.params.uniform, code);
   });
 
+function generatePointerCheck(check: string): string {
+  if (check === `address`) {
+    return `let tmp = workgroupUniformLoad(ptr);`;
+  } else {
+    // check === `contents`
+    return `if test_val > 0 {
+      workgroupBarrier();
+    }`;
+  }
+}
+
+const kPointerCases = {
+  address_uniform_literal: {
+    code: `let ptr = &wg_array[0];`,
+    check: `address`,
+    uniform: true,
+  },
+  address_uniform_value: {
+    code: `let ptr = &wg_array[uniform_value];`,
+    check: `address`,
+    uniform: true,
+  },
+  address_nonuniform_value: {
+    code: `let ptr = &wg_array[nonuniform_value];`,
+    check: `address`,
+    uniform: false,
+  },
+  address_uniform_chain: {
+    code: `let p1 = &wg_struct.x;
+    let p2 = &(*p1)[uniform_value];
+    let p3 = &(*p2).x;
+    let ptr = &(*p3)[uniform_value];`,
+    check: `address`,
+    uniform: true,
+  },
+  address_nonuniform_chain1: {
+    code: `let p1 = &wg_struct.x;
+    let p2 = &(*p1)[nonuniform_value];
+    let p3 = &(*p2).x;
+    let ptr = &(*p3)[uniform_value];`,
+    check: `address`,
+    uniform: false,
+  },
+  address_nonuniform_chain2: {
+    code: `let p1 = &wg_struct.x;
+    let p2 = &(*p1)[uniform_value];
+    let p3 = &(*p2).x;
+    let ptr = &(*p3)[nonuniform_value];`,
+    check: `address`,
+    uniform: false,
+  },
+  wg_uniform_load_is_uniform: {
+    code: `let test_val = workgroupUniformLoad(&wg_scalar);`,
+    check: `contents`,
+    uniform: true,
+  },
+  contents_scalar_uniform: {
+    code: `let ptr = &func_scalar;
+    let test_val = *ptr;`,
+    check: `contents`,
+    uniform: true,
+  },
+  contents_scalar_nonuniform1: {
+    code: `func_scalar = nonuniform_value;
+    let ptr = &func_scalar;
+    let test_val = *ptr;`,
+    check: `contents`,
+    uniform: false,
+  },
+  contents_scalar_nonuniform2: {
+    code: `let ptr = &func_scalar;
+    *ptr = nonuniform_value;
+    let test_val = *ptr;`,
+    check: `contents`,
+    uniform: false,
+  },
```

Comment:
Would love to see a couple more slight variations of these. One where the non-uniform assignment happens on the original variable *after* you've captured the pointer:
```
let ptr = &func_scalar;
func_scalar = nonuniform_value;
let test_val = *ptr;
```

And then another where you assign a non-uniform through the pointer but then test the original variable:
```
let ptr = &func_scalar;
*ptr = nonuniform_value;
let test_val = func_scalar;
```

These were some corner cases that tripped up my initial implementation in Tint.

Reply:
Added these in contents_scalar_nonuniform 3 through 5.

Reply:
Looks great, thanks! 

If it's not too much trouble it might also be worth adding a couple that test the opposite, where a value starts as non-uniform and then becomes uniform:
e.g.
```
func_scalar = non_uniform_value;
let ptr = &func_scalar;
func_scalar = uniform_value;
let test_val = *ptr;
```
etc

Reply:
Added in most recent commit